### PR TITLE
[PoC] In-place struct initialization

### PR DESCRIPTION
**Pull request by wilzbach**
So this is my WIP hack to have in-place struct initialization into DMD.
As during the parsing stage we don't have much information about the struct and can only parse a `StructInitializer`, I created an intermediate `StructLiteralExp2` that will be converted into a `StructLiteralExp` during expression semantic.
I would like to have a more elegant solution and this still needs a lot of work, but it already passes the test suite for the common cases of this DIP and I thought it's good to show that it's possible to do this in dmd for submitting the DIP. Also I get to see what the CIs say.

What's in-place struct initialization?
-------------------------------------------

Full DIP: https://github.com/dlang/DIPs/pull/71

tl;dr:

```d
struct StructFoo
{
    int a;
}

void bar(StructFoo foo)
{
    assert(foo.a == 3);
}

StructFoo f = {a: 1};
f = StructFoo({a: 2});
assert(f.a == 2);
bar(StructFoo({a: 3}));
```

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Why do you need a new expression type?

**Comment by RazvanN7** [comment]
Can't you just use the existing StructLiteralExp and add the fields fields you want + a bool field that tells you if StructLiteralExp is used for in-place initialization?

**Comment by andralex** [comment]
status?

**Comment by jacob-carlborg** [comment]
> status?

@andralex there's a DIP that hasn't been approved. Personally I don't think a DIP is needed. But we seem to disagree on the syntax.

**Comment by wilzbach** [comment]
> status?

This is a working PoC, but there are multiple ways to integrate this feature in the D grammar (see the [DIP](https://github.com/dlang/DIPs/pull/71) for more details) and as there was never a consensus on which syntax to use. @andralex what are your thoughts on the syntax for this?
Once that's decided, polishing & merging this would be easy...

**Comment by dkorpel** [comment]
This has been superseded by named arguments, which have been implemented for struct constructors.
